sdks/java/core/src/test/java/org/apache/beam/sdk/io/LocalFileSystemTest.java

```diff
@@ -147,6 +153,122 @@ private void testCopy(List<Path> srcFiles, List<Path> destFiles, List<String> co
+    List<MatchResult> matchResults = localFileSystem.match(
+        ImmutableList.of(temporaryFolder.getRoot().toPath().resolve("b") + "*"));
+    assertThat(
+        toFilenames(matchResults),
+        containsInAnyOrder(expected.toArray(new String[expected.size()])));
+  }
+
+  @Test
+  public void testMatchUsingExplicitPath() throws Exception {
+    List<String> expected = ImmutableList.of(temporaryFolder.newFile("a").toString());
+    temporaryFolder.newFile("aa");
+
+    List<MatchResult> matchResults = localFileSystem.match(
+        ImmutableList.of(temporaryFolder.getRoot().toPath().resolve("a").toString()));
+    assertThat(
+        toFilenames(matchResults),
+        containsInAnyOrder(expected.toArray(new String[expected.size()])));
+  }
+
+  @Test
+  public void testMatchUsingExplicitPathForNonExistentFile() throws Exception {
+    List<String> expected = ImmutableList.of();
+    temporaryFolder.newFile("aa");
+
+    List<MatchResult> matchResults = localFileSystem.match(
+        ImmutableList.of(temporaryFolder.getRoot().toPath().resolve("a").toString()));
+    assertThat(
+        toFilenames(matchResults),
+        containsInAnyOrder(expected.toArray(new String[expected.size()])));
+  }
+
+  @Test
+  public void testMatchMultipleWithoutSubdirectoryExpansion() throws Exception {
+    File unmatchedSubDir = temporaryFolder.newFolder("aaa");
+    File unmatchedSubDirFile = File.createTempFile("sub-dir-file", "", unmatchedSubDir);
+    unmatchedSubDirFile.deleteOnExit();
+    List<String> expected = ImmutableList.of(temporaryFolder.newFile("a").toString(),
+        temporaryFolder.newFile("aa").toString(), temporaryFolder.newFile("ab").toString());
+    temporaryFolder.newFile("ba");
+    temporaryFolder.newFile("bb");
+
+    // Windows doesn't like resolving paths with * in them, so the * is appended after resolve.
+    List<MatchResult> matchResults = localFileSystem.match(
+        ImmutableList.of(temporaryFolder.getRoot().toPath().resolve("a") + "*"));
+    assertThat(
+        toFilenames(matchResults),
+        containsInAnyOrder(expected.toArray(new String[expected.size()])));
+  }
+
+  @Test
+  public void testMatchMultipleWithSubdirectoryExpansion() throws Exception {
+    File matchedSubDir = temporaryFolder.newFolder("a");
+    File matchedSubDirFile = File.createTempFile("sub-dir-file", "", matchedSubDir);
+    matchedSubDirFile.deleteOnExit();
+    File unmatchedSubDir = temporaryFolder.newFolder("b");
+    File unmatchedSubDirFile = File.createTempFile("sub-dir-file", "", unmatchedSubDir);
+    unmatchedSubDirFile.deleteOnExit();
+
+    List<String> expected = ImmutableList.of(matchedSubDirFile.toString(),
+        temporaryFolder.newFile("aa").toString(), temporaryFolder.newFile("ab").toString());
+    temporaryFolder.newFile("ba");
+    temporaryFolder.newFile("bb");
+
+    // Windows doesn't like resolving paths with * in them, so the ** is appended after resolve.
```

Comment:
Separate this windows-specific logic into one place where it can be cleanly reviewed, then use that utility function throughout.